src/backend/stream.rs

```diff
@@ -0,0 +1,821 @@
+{
+  let stm = &*(u as *mut Stream);
+  pa_threaded_mainloop_signal(stm.context.mainloop, 0);
+}
+
+unsafe extern fn stream_drain_callback(a: *mut pa_mainloop_api, e: *mut pa_time_event, _tv: *const Struct_timeval, u: *mut c_void)
+{
+  let mut stm = &mut *(u as *mut Stream);
+  debug_assert!(stm.drain_timer == e);
+  stm.state_change_callback(cubeb::State::Drained);
+  /* there's no pa_rttime_free, so use this instead. */
+  (*a).time_free.unwrap()(stm.drain_timer);
+  stm.drain_timer = ptr::null_mut();
+  pa_threaded_mainloop_signal(stm.context.mainloop, 0);
+}
+
+unsafe extern fn stream_state_callback(s: *mut pa_stream,
+                                       u: *mut c_void)
+{
+    let stm = &mut *(u as *mut Stream);
+    if !PA_STREAM_IS_GOOD(pa_stream_get_state(s)) {
+        stm.state_change_callback(cubeb::State::Error);
+    }
+    pa_threaded_mainloop_signal(stm.context.mainloop, 0);
+}
+
+
+fn read_from_input(s: *mut pa_stream, buffer: *mut *const c_void, size: *mut usize) -> i32
+{
+  let readable_size = pa_stream_readable_size(s);
+  if readable_size > 0 {
+    if pa_stream_peek(s, buffer, size) < 0 {
+      return -1;
+    }
+  }
+
+  readable_size as i32
+}
+
+unsafe extern fn stream_write_callback(s: *mut pa_stream, nbytes: usize, u: *mut c_void)
+{
+//  LOGV("Output callback to be written buffer size %zd", nbytes);
+    let mut stm = &mut *(u as *mut Stream);
+    if stm.shutdown || stm.state != cubeb::State::Started {
+        return;
+    }
+
+    if stm.input_stream.is_null() {
+        // Output/playback only operation.
+        // Write directly to output
+        debug_assert!(!!stm.input_stream.is_null() && !stm.output_stream.is_null());
+        stm.trigger_user_callback(s, ptr::null(), nbytes);
+    }
+}
+
+unsafe extern fn stream_read_callback(s: *mut pa_stream, _nbytes: usize, u: *mut c_void)
+{
+    //  LOGV("Input callback buffer size %zd", nbytes);
+    let mut stm = &mut *(u as *mut Stream);
+    if stm.shutdown {
+        return;
+    }
+
+    let mut read_data: *const c_void = ptr::null();
+    let mut read_size: usize = 0;
+    while read_from_input(s, &mut read_data, &mut read_size) > 0 {
+        /* read_data can be NULL in case of a hole. */
+        if !read_data.is_null() {
+            let in_frame_size = pa_frame_size(&stm.input_sample_spec);
+            let read_frames = read_size / in_frame_size;
+
+            if !stm.output_stream.is_null() {
+                // input/capture + output/playback operation
+                let out_frame_size = pa_frame_size(&stm.output_sample_spec);
+                let write_size = read_frames * out_frame_size;
+                // Offer full duplex data for writing
+                let stream = stm.output_stream;
+                stm.trigger_user_callback(stream, read_data, write_size);
+            } else {
+                // input/capture only operation. Call callback directly
+                let got = stm.data_callback.unwrap()(stm as *const _ as *mut _,
```

Comment:
You're right. It should be stm as *mut _ as * mut _. First to convert &mut Stream to *mut Stream and second to convert *mut Stream to *mut c_void.